javaslang/src/test/java/javaslang/MatchTest.java

```diff
@@ -219,6 +219,26 @@ public void shouldDecomposeSomeSomeTuple() {
 
     // -- List
 
+    @Test
+    public void shouldDecomposeEmptyList() {
```

Comment:
The match pattern definitions (see $.java) have to return a Tuple. Therefore using isEmpty is not possible (the method returns either `Tuple0` or `Tuple2` but it cannot return both).

Providing separate classes is the only possibility I see. Then my dream would come true and all collections could be interfaces with two nested classes `NonEmpty` and `Empty`. I like that idea. Those collections without an ordering/Comparator could have a singleton Empty class.

(Having interfaces is nice because we are able to proxy them. I make use of that in `Lazy.val` for example. But the nested generic `T` in `A<T>` is nasty.)